src/impl-genera.lisp

```diff
@@ -38,41 +47,56 @@ Distributed under the MIT license (see LICENSE file)
   (check-type lock lock)
   (setf (lock-lock-argument lock) (process:make-lock-argument (lock-lock lock)))
   (if wait-p
```

Comment:
Use a COND here.

Reply:
I just committed the changes you requested above.